core/src/main/java/io/questdb/cutlass/http/client/HttpClient.java

```diff
@@ -0,0 +1,463 @@
+
+import io.questdb.HttpClientConfiguration;
+import io.questdb.cutlass.http.HttpHeaderParser;
+import io.questdb.network.IOOperation;
+import io.questdb.network.NetworkFacade;
+import io.questdb.std.*;
+import io.questdb.std.str.AbstractCharSink;
+import io.questdb.std.str.CharSink;
+import io.questdb.std.str.DirectByteCharSequence;
+import io.questdb.std.str.StringSink;
+
+public abstract class HttpClient implements QuietCloseable {
+    protected final NetworkFacade nf;
+    private final int bufferSize;
+    private final ObjectPool<DirectByteCharSequence> csPool = new ObjectPool<>(DirectByteCharSequence.FACTORY, 64);
+    private final int defaultTimeout;
+    private final Request request = new Request();
+    protected int fd = -1;
+    private long bufLo;
+    private long ptr = bufLo;
+    private ResponseHeaders responseHeaders;
+
+    public HttpClient(HttpClientConfiguration configuration) {
+        this.nf = configuration.getNetworkFacade();
+        this.defaultTimeout = configuration.getTimeout();
+        this.bufferSize = configuration.getBufferSize();
+        this.bufLo = Unsafe.malloc(bufferSize, MemoryTag.NATIVE_DEFAULT);
+        this.responseHeaders = new ResponseHeaders(bufLo, bufferSize, defaultTimeout, 4096, csPool);
+    }
+
+    @Override
+    public void close() {
+        disconnect();
+        if (bufLo != 0) {
+            Unsafe.free(bufLo, bufferSize, MemoryTag.NATIVE_DEFAULT);
+            bufLo = 0;
+        }
+        responseHeaders = Misc.free(responseHeaders);
+    }
+
+    public void disconnect() {
+        if (fd != -1) {
+            nf.close(fd);
+            fd = -1;
+        }
+    }
+
+    public Request newRequest() {
+        ptr = bufLo;
+        request.state = Request.STATE_REQUEST;
+        return request;
+    }
+
+    private int die(int byteCount) {
+        if (byteCount < 1) {
+            throw new HttpClientException("peer disconnect [errno=").errno(nf.errno()).put(']');
+        }
+        return byteCount;
+    }
+
+    private int recvOrDie(long addr, int timeout) {
+        return recvOrDie(addr, (int) (bufferSize - (addr - bufLo)), timeout);
+    }
+
+    private int recvOrDie(long lo, int len, int timeout) {
+        ioWait(timeout, IOOperation.READ);
+        return die(nf.recv(fd, lo, len));
+    }
+
+    private int sendOrDie(long lo, int len, int timeout) {
+        ioWait(timeout, IOOperation.WRITE);
+        return die(nf.send(fd, lo, len));
+    }
+
+    protected void dieWaiting(int n) {
+        if (n == 1) {
+            return;
+        }
+
+        if (n == 0) {
+            throw new HttpClientException("timed out [errno=").put(nf.errno()).put(']');
+        }
+
+        throw new HttpClientException("queue error [errno=").put(nf.errno()).put(']');
+    }
+
+    protected abstract void ioWait(int timeout, int op);
+
+    protected abstract void setupIoWait();
+
+    private static class BinarySequenceAdaptor implements BinarySequence, Mutable {
```

Comment:
nit: in all other classes we use `Adapter`, not `Adaptor`

Reply:
done in https://github.com/questdb/questdb/pull/3399/commits/52e369ab6b2e99ce2b12ebcc9b1c558e75dd0716